tiledb/sm/array_schema/tile_domain.h

```diff
@@ -0,0 +1,356 @@
+   * tile domain of the instance.
+   */
+  bool in_tile_domain(const T* tile_coords) const {
+    for (unsigned i = 0; i < dim_num_; ++i) {
+      if (tile_coords[i] < tile_domain_[2 * i] ||
+          tile_coords[i] > tile_domain_[2 * i + 1])
+        return false;
+    }
+
+    return true;
+  }
+
+  /**
+   * Returns the global subarray corresponding to the tile with
+   * the input coordinates.
+   */
+  std::vector<T> tile_subarray(const T* tile_coords) const {
+    std::vector<T> ret;
+    ret.resize(2 * dim_num_);
+
+    for (unsigned i = 0; i < dim_num_; ++i) {
+      ret[2 * i] = tile_coords[i] * tile_extents_[i] + domain_[2 * i];
+      ret[2 * i + 1] = ret[2 * i] + tile_extents_[i] - 1;
+    }
+
+    return ret;
+  }
+
+  /**
+   * Returns the tile overlap (as a global subarray slice) between the
+   * domain slice of the instance and the tile identified by `tile_coords`.
+   *
+   * If there is no overlap, then the returned vector is empty.
+   */
+  std::vector<T> tile_overlap(const T* tile_coords) const {
+    std::vector<T> ret;
+
+    // Return empty if the tile is not in the tile domain
+    if (!in_tile_domain(tile_coords))
+      return ret;
+
+    // Get overlap
+    ret.resize(2 * dim_num_);
+    auto tile_subarray = this->tile_subarray(tile_coords);
+    for (unsigned i = 0; i < dim_num_; ++i) {
+      ret[2 * i] = MAX(tile_subarray[2 * i], domain_slice_[2 * i]);
+      ret[2 * i + 1] = MIN(tile_subarray[2 * i + 1], domain_slice_[2 * i + 1]);
+    }
+
+    return ret;
+  }
+
+  /**
+   * Returns `true` if the instance's domain slice covers completely
+   * that of `tile_domain` for the tile identified by `tile_coords`.
+   */
+  bool covers(const T* tile_coords, const TileDomain& tile_domain) const {
+    if (!in_tile_domain(tile_coords) ||
+        !tile_domain.in_tile_domain(tile_coords))
+      return false;
+
+    auto tile_overlap_1 = this->tile_overlap(tile_coords);
+    auto tile_overlap_2 = tile_domain.tile_overlap(tile_coords);
+    assert(tile_overlap_1.size() == tile_overlap_2.size());
+    assert(tile_overlap_1.size() == 2 * dim_num_);
+    for (unsigned i = 0; i < dim_num_; ++i) {
+      if (tile_overlap_2[2 * i] < tile_overlap_1[2 * i] ||
+          tile_overlap_2[2 * i + 1] > tile_overlap_1[2 * i + 1])
+        return false;
+    }
+
+    return true;
+  }
+
+  /**
+   * Given the input tile coordinates, this function produces an
+   * 1D-mapped tile position, based on the layout of the tile
+   * domain. Note that the returned tile position will be normalized
+   * to reflect the position within the tile domain. That is,
+   * if the tile (say 1D)domain is `[5,8]`, and the tile coords
+   * are `6`, then the returned position will be `1` (starting
+   * always from 0).
+   *
+   * If the input tile coordinates are not inside the tile domain,
+   * the function returns `UINT64_MAX` (for invalid position).
+   */
+  uint64_t tile_pos(const T* tile_coords) const {
```

Comment:
what happens if the tile domain coordinates include UNIT64_MAX in the original domain?

Reply:
I guess my question is if there is a possibility of pos == UINT64_MAX or overflowing

Reply:
No this is not a problem mainly because of this check: https://github.com/TileDB-Inc/TileDB/blob/dev/tiledb/sm/array_schema/dimension.cc#L422